### PR TITLE
Semi-automatic derivation with Magnolia

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Addresses #135, #201 and perhaps #168 too.

This PR substitutes [Magnolia](http://magnolia.work/) for Shapeless in the derivation of the `DynamoFormat` and `EnumerationDynamoFormat` typeclasses. It's still rough around the edges but it's getting there. To derive a format, users must now write, for type `T`:
```scala
implicit val format: DynamoFormat[T] = DynamoFormat.derive // if T is not an enumeration
implicit val format: EnumerationDynamoFormat[T] = DynamoFormat.deriveEnum // otherwise
```

I'm not familiar enough with the [export-hook](https://github.com/milessabin/export-hook) thingie so will need to do some reading.

TODO
- [x] Fix tests
- [x] Update scaladocs
- [x] Update tuts

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
- I think automatic derivation should still be available somewhere, ala Circe
- Regarding package organisation, perhaps offering package objects `semiauto` and `auto` then
- I'm wondering if the enumeration derivation is still necessary, both `combine` and `dispatch` seem general enough

**Comment by philwills** [comment]
> I think automatic derivation should still be available somewhere, ala Circe
👍 

> I'm wondering if the enumeration derivation is still necessary, both combine and dispatch seem general enough
This needs a lot of care, as I had thought that previously and ran into #138. I haven't looked at this carefully enough to know, but it's worth bearing in mind.

**Comment by philwills** [comment]
It would be great to know whether there's a noticeable compilation speed-up for a deeply ADT.

In the documentation, I'd err towards putting the derived format in the companion object, as instinctively that would be my preferred pattern. Happy to be argued out of that though.

**Comment by regiskuckaertz** [comment]
> This needs a lot of care, as I had thought that previously and ran into #138.

Yeah, I think I found a way. Magnolia provides an `isObject` method on `CaseClass` and it provides a nice solution to the problem.

Now two tests are failing and I'm not sure why yet.

File: alpakka/src/test/scala/com/gu/scanamo/ScanamoAlpakkaSpec.scala [hunk 1]
```diff
@@ -145,7 +147,7 @@ class ScanamoAlpakkaSpec
       import com.gu.scanamo.syntax._
 
       val forecasts = for {
-        _ <- ScanamoAlpakka.update(alpakkaClient)("forecast")('location -> "London", set('weather -> "Sun"))
+        _ <- ScanamoAlpakka.update[Forecast](alpakkaClient)("forecast")('location -> "London", set('weather -> "Sun"))
```
Comment (regiskuckaertz) [inline alpakka/src/test/scala/com/gu/scanamo/ScanamoAlpakkaSpec.scala]: Not adding the type generates a diverging implicit error, but I see this is used in other places so assuming that's not too bad.

**Comment by regiskuckaertz** [comment]
Automatic derivation for any `A` supplanted the derivation for `Some[A]` in `DynamoFormat`. That's where the export-hook thing comes in... is this a caveat we're willing to add to the documentation? Speaking of, I need to update the pages on custom formats...

Otherwise, looking good?

File: alpakka/src/test/scala/com/gu/scanamo/ScanamoAlpakkaSpec.scala (hunk 1 above)
Comment (philwills) [reply alpakka/src/test/scala/com/gu/scanamo/ScanamoAlpakkaSpec.scala]: I'm OK with it in this case. I'd be a lot more worried if it was one of the methods on `Table`. Just need to think about how to make it clear in release notes.

**Comment by philwills** [comment]
This does look really good generally. I'm not quite sure what you mean with `Some[T]`, but `Option` is such a common type that any regression in it's ergonomics would be an issue.

**Comment by regiskuckaertz** [comment]
> I'm not quite sure what you mean with Some[T]

The failing tests occurred when a column was of type `Option[T]` and the program tried to put `Some(x)` into it. There is a rule in the `DynamoFormat` object for that specific case, but it was foreshadowed by bringing `generic.auto._` into scope.

**Comment by philwills** [comment]
Got it. Having the auto-derivation as part of the class hierarchy made it relatively simple to push it's priority down the list, but having it imported trumps the precedence of the companion object.

My usual tactic in these situations is to look what Circe has done and blatantly rip it off. https://github.com/circe/circe/pull/247/files looks like the most relevant diff, but there's a lot going on there to pick out the relevant parts from.

**Comment by regiskuckaertz** [comment]
haha, good ideas travel far in the open source world 😄 

ok I see what he's doing, though I don't understand yet how it's working wrt priority. I'll try and see how I can reconcile this with Magnolia's macro. My first attempt (just wrapping an instance in `Exported` at the calling site) did not work.

**Comment by regiskuckaertz** [comment]
oh I get it. The implicit def in `LowPriority` expects an implicit `Exported[DynamoFormat[A]]`, but that can only exist if you bring `io.circe.generic.auto._` in scope where it's being defined.

**Comment by fommil** [comment]
great! Please add a `deriving.conf` using this format https://gitlab.com/fommil/scalaz-deriving/#deriving so that it works out of the box with `@deriving`

e.g. https://gitlab.com/fommil/scalaz-deriving/blob/master/examples/xmlformat/src/main/resources/deriving.conf

You could even add unit tests for this.

**Comment by regiskuckaertz** [comment]
@fommil ooh nice 👍just did.

**Comment by regiskuckaertz** [comment]
@philwills there's current activity in the magnolia project on that very issue (automatic derivation + scope), hopefully we can get that sorted soon.

**Comment by dmitry-avershin** [comment]
what is the state of this one? would it be part of the upcoming release?

**Comment by regiskuckaertz** [comment]
No, it won't go in, because fully automatic derivation subsumes implicits for concrete types, which I haven't found a solution to yet.
We're working on semi+full auto derivation within the existing shapeless code and that is going to ship in v1.

**Comment by regiskuckaertz** [comment]
I expect scala 3 will be out before the auto derivation issue is solved

**Pull request closed without merging**